frameworks/Rust/hyper/src/main.rs

```diff
@@ -1,91 +1,211 @@
+}
+
+async fn serve(handle: &runtime::Handle) -> Result<()> {
+    let addr = SocketAddr::from((Ipv4Addr::UNSPECIFIED, 8080));
+    let socket = create_socket(addr)?;
+
+    let listener = TcpListener::from_std(socket.into())?;
+    let addr = listener.local_addr()?;
+    info!("Listening on: {}", addr);
+
+    // spawn accept loop into a task so it is scheduled on the runtime with all the other tasks.
+    let accept_loop = accept_loop(handle.clone(), listener);
+    handle.spawn(accept_loop).await.unwrap()
+}
+
+/// Create a socket that allows reuse of the address and port.
+///
+/// This makes it possible for multiple instances of the server task to bind to the same address and
+/// port.
+fn create_socket(addr: SocketAddr) -> Result<Socket> {
+    let domain = match addr {
+        SocketAddr::V4(_) => Domain::IPV4,
+        SocketAddr::V6(_) => Domain::IPV6,
+    };
+    let addr = SockAddr::from(addr);
+    let socket = Socket::new(domain, socket2::Type::STREAM, None)?;
+    let backlog = 4096; // maximum number of pending connections
+    #[cfg(unix)]
+    socket.set_reuse_port(true)?;
+    socket.set_reuse_address(true)?;
+    socket.set_nonblocking(true)?; // required for tokio
+    socket.bind(&addr)?;
+    socket.listen(backlog)?;
+
+    Ok(socket)
+}
+
+/// Accept loop that accepts incoming connections and spawns a new task to handle each connection.
+async fn accept_loop(handle: runtime::Handle, listener: TcpListener) -> Result<()> {
+    let mut http = http1::Builder::new();
+    http.pipeline_flush(true);
+
+    let service = service_fn(router);
+    loop {
+        let (stream, _) = listener.accept().await?;
+        let http = http.clone();
+        handle.spawn(async move {
+            let io = TokioIo::new(stream);
+            if let Err(_e) = http.serve_connection(io, service).await {
+                // ignore errors until https://github.com/hyperium/hyper/pull/3863/ is merged
+                // This PR will allow us to filter out shutdown errors which are expected.
+                // warn!("Connection error (this may be normal during shutdown): {e}");
             }
         });
+    }
+}
+
+/// Routes requests to the appropriate handler.
+async fn router(request: Request<Incoming>) -> Result<Response<BoxBody<Bytes, Infallible>>> {
+    // The method is always GET, so we don't check it.
+    match request.uri().path() {
+        "/ping" => ping(),
+        "/json" => json::get(),
+        "/db" => single_query::get().await,
+        "/queries" => multiple_queries::get(request.uri().query()).await,
+        "/fortunes" => fortunes::get().await,
+        "/plaintext" => plaintext::get(),
+        _ => not_found_error(),
+    }
+}
+
+/// A handler that returns a "pong" response.
+///
+/// This handler is used to verify that the server is running and can respond to requests. It is
+/// used by the docker health check command.
+fn ping() -> Result<Response<BoxBody<Bytes, Infallible>>> {
```

Comment:
I had an incorrectly configured test and was struggling to get the benchmarks to properly connect to that one test. I wanted a way to ensure that the server was known to be up outside of the tests themselves, so added this to support the docker healthchecks.

(The underlying problem that I was hitting was caused by me returning an overly large static value for content length from the db test - 32 instead of actually calculating the length. This causes the test framework to not be able to connect to that test and also to not report which test is failing, only that it can't connect)